### DISCUSSION TITLE
Snippets for classes

### DISCUSSION THREAD

**Discussion by PhosCity** [L3MON4D3/LuaSnip#724] [Q&A] [answered]
Hello! Just diving into the snippets and learned the basics. Thank you for the amazing plugin!

I have a question. How do I make snippets for different methods of a class. For example let's say I have a class named `data` and it has different methods like `parse`, `commit`, `get` etc that I can access in the following manner:

```moonscript
data\parse!
data\get!
data\commit!
```

So when I type `data\` I want to choose from it's various methods and then execute the snippet for the particular method I choose. I can make a choice node for choosing differnt methods of the class but I don't think choice nodes is the answer here since each method have their own snippets that I want to execute.

How difficult is this to achieve? Is there an easy way that even a beginner me can do?

**Comment by atticus-sullivan** [comment]
Hm, to be honest this doesn't sound quite like an application for snippets. More like something which should/can be achieved by using a Language Server. This will also (try to) track of what type a variable is and can therefore suggest the right methods for it. At least the latter is out of scope for a snippet plugin.

For neovim, you can easily use https://github.com/neovim/nvim-lspconfig (configures the by now in nvim builtin language server client) in combination with the right language server and something like https://github.com/hrsh7th/nvim-cmp (for the completion menu) this should work. 

If you want to stick with doing this via snippets (the string `data` has to be kinda fixed for this), you could just write the snippets separately (no choice nodes) with `data\` as prefix for the trigger. Your completion menu plugin should just show you all expandable snippets (which should be only or at least mainly the ones you configured) when you type `data\`

**Reply by PhosCity** [reply]
The language I want to write the snippet is Moonscript which afaik does not have language servers or treesitter grammar or anything else. So I unfortunately can't use them. That's why I'm writing snippets for it.

If I understand the last part correctly,  I should create multiple snippets with triggers like `data\parse`, `data\commit`, `data\get` etc right?

**Reply by atticus-sullivan** [reply]
Ah, I see (and you also can't simply use the lua language server, since it is quite different).

Yes that's exactly what I would do (but maybe others here have other suggestions).

For some sort of typing sensitiveness, you could try to embed the type in the variable name and then use some lua pattern matching for the trigger. Right now  the snippets don't show in cmp, but this should change in the future (https://github.com/saadparwaiz1/cmp_luasnip/pull/46).
But of course that's not an ideal solution.

**Comment by L3MON4D3** [comment, marked as answer]
Oh, this could be done by writing a custom `nvim-cmp`-source (actually not that hard!!). Probably way too overkill, but it seems like a perfect solution.
But for practical purposes, using snippets with the same prefix, as @atticus-sullivan suggests, makes more sense since it will most likely work just as well with the regular snippet-completion from `cmp_luasnip`

Alternatively, you can use [`condition` and `show_condition`](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snippets) to check whether the characters before the trigger match `"data\"`, and then allow or prevent expansion/completion.

**Reply by PhosCity** [reply]
Wow thanks for letting me know about `condition` and `show_condition`. Is the following snippet proper usage of it?

```lua
  s(
    { trig = "data\\callback", dscr = "callback function for each line section" },
    fmt(
      [[
      data\callback ({}) ->
        {}
      ]],
      {
        c(1, { t("section"), t("sections, section, i") }),
        i(0),
      }
    ),
    {
      show_condition = function(line_to_cursor)
        if line_to_cursor == "data\\" then
          return true
        end
      end,
    }
  ),
```

If it's not, could you give me an example of it?

**Reply by L3MON4D3** [reply]
That looks correct👍
You might want to only check the last few characters of line_to_cursor, and maybe use the same function in condition, since that will also prevent expansion via trigger.
Does it work?

**Reply by PhosCity** [reply]
Yup seems to be working. I was asking if the format was incorrect.

You're right about only checking the last few characers because the snippet I posted above would not work if the line was indented or had anything else before it. Here's the final version.

```lua
show_condition = function(line_to_cursor)
  return string.sub(line_to_cursor, -5):match("data\\")
end,
```

Thank you for all the help

**Reply by L3MON4D3** [reply]
Ah, okay :)
Looks nice :+1:
